SQL-1058: Implement SQLGetInfoW

mattChiaravalloti opened the pull request:
This PR implements `SQLGetInfoW`. This is a work in progress but is ready for review since the remaining work requires help! I've left several comments in the code labeled `"TODO: (code review)"` where I'm looking for some input on what to do/how to handle certain values.

This PR also does not yet have any tests, but I'm wondering what you think about that. Should these be unit tested? Nearly all of them return constant values, so there is not really any special computation to test. If we ever need to change the behavior, we'll end up changing the tests in exactly the same way: expecting the new constant result. Let me know if you have ideas on testing for this!

mattChiaravalloti commented on odbc/src/api/data.rs:
```diff
@@ -408,6 +408,30 @@ pub mod i16_len {
+    /// # Safety
+    /// This writes to multiple raw C-pointers
+    ///
+    pub unsafe fn set_output_fixed_data<T: core::fmt::Debug>(
```
We needed to add this one to the `i16_len` module since `SQLGetInfoW` uses a `*mut SmallInt` for the length.

mattChiaravalloti commented on odbc/src/api/definitions.rs:
```diff
@@ -110,3 +110,256 @@ pub enum UseBookmarks {
+
+#[repr(u16)]
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+pub enum InfoType {
```
The `odbc_sys::InfoType` enum does not contain many of the variants we need. I created this enum with exactly the set of `InfoType`s we need to support for now.

mattChiaravalloti commented on odbc/src/api/definitions.rs:
```diff
@@ -110,3 +110,256 @@ pub enum UseBookmarks {
+pub const SQL_INFO_Y: &str = "Y";
+
+// SQL_CONVERT_FUNCTIONS bitmask
+pub const SQL_FN_CVT_CAST: u32 = 0x00000002;
```
In general here and below I only include the values that we actually use. The alternative would be including _all_ of the values for a certain set (for example, here, the `SQL_FN_CVT_CONVERT: u32 = 0x00000001;` value) and marking it with a rust attribute to ignore dead/unused code. Not sure that's worth it though! I think this current implementation is cleaner since it avoids dead/unused code.

mattChiaravalloti commented on odbc/src/api/functions.rs:
```diff
@@ -1785,11 +1785,441 @@ pub unsafe extern "C" fn SQLGetInfo(
+        // SQL_DRIVER_NAME
+        InfoType::DriverName => {
+            // A character string with the file name of the driver used to access the data source.
+            // TODO: (code review) What file name should be used here? Is this something that
+            //       will need to be dynamically determined or is it something we can hardcode? 
```
This is the style of comment I'm talking about: in these `TODO: (code review)` comments, I added some info about the `InfoType` and asked questions. I appreciate any help!

nbagnard replied:
I checked with Alexi.
To match the JDBC driver, she wants to use : MongoDB Atlas SQL interface ODBC Driver

nbagnard commented on odbc/src/api/functions.rs:
```diff
@@ -1785,11 +1785,441 @@ pub unsafe extern "C" fn SQLGetInfo(
+            return i16_len::set_output_fixed_data(&SQL_CU_DML_STATEMENTS, info_value_ptr, string_length_ptr);
+        }
+        // SQL_SPECIAL_CHARACTERS
+        InfoType::SpecialCharacters => {
```
The important thing here is **when do we require the identifier to be a quoted identifier.**
Postgres returns only "_".
I'd say that we need the identifier to be quoted if it contains `, ", ', . or $ (the last 2 are not supported yet), no?
Any other character that you see would throw the parser or algebrizer off?  I so, add it to this list.

mattChiaravalloti replied:
The [spec says](https://docs.google.com/document/d/14DnyMSzVNRkoyib8yM62MPJIrkrG1pE0KjRYOWOUuDc/edit#heading=h.ea137t3xw581) that regular (unquoted) identifiers can only consist of ascii letters, digits, and underscore, and everything else must be quoted:
```
<regular identifier> ::= ([A-Za-z] | "_")[A-Za-z0-9_]*

<delimited identifier> ::= " <identifier character>* "
                         | ` <identifier character>* `

<identifier character> ::= [^\x00]
```
Plus we have spec tests asserting this behavior, so to match your suggestion would require a spec change. We can go down that route, but we may want to discuss this as a team before committing to this.

nbagnard replied:
umh, I'm not sure why you'd require a spec change?
It seems more like this info type has a very bad format and it is not realistic to return a string with all the possible character that we accept. Returning nothing does not seem that great either, because it implies that we don't need the quotes for any characters.
I would love to be able to provide a regex but if application expects a list of chars to know if they need to quote identifier or not, then it won't work.
Covering the basic most used chars is maybe not 100% better because then it is incorrect too, but it seems more usable.
Ultimately, I believe most app always put identifier in quotes to be on the safe side.
Let's talk about this a bit more tomorrow

mattChiaravalloti replied:
Updated the implementation to return a string containing some special characters like you brought up. lmk what you think!

nbagnard replied:
I think that's already quite a lot and will probably go a long way 👍 Thank you for taking the time to think about useful ones :)

nbagnard commented on odbc/src/api/functions.rs:
```diff
@@ -1785,11 +1785,441 @@ pub unsafe extern "C" fn SQLGetInfo(
+            //  - MongoSQL's grammar defines an identifier character as [^\x00], as in, literally any non-null character.
+            //    Surely we aren't meant to return a "special characters" string that contains every character, are we?
+        }
+        // SQL_MAX_COLUMNS_IN_GROUP_BY
```
[nit] You could bundle all the MaxColumnsXXX info type cases since there is no max for any of them.

mattChiaravalloti replied:
done

nbagnard commented on odbc/src/api/functions.rs:
```diff
@@ -1785,11 +1785,441 @@ pub unsafe extern "C" fn SQLGetInfo(
+            // columns allowed in a SELECT clause.
+            return i16_len::set_output_fixed_data(&SQL_U16_ZERO, info_value_ptr, string_length_ptr);
+        }
+        // SQL_TIMEDATE_ADD_INTERVALS
```
Note: Until we support TIMESTAMPADD and TIMESTAMPDIFF this will not be used, but it's good to have it ready.

mattChiaravalloti replied:
Left this implemented but updated the comment to reflect it will not be used. Should I actually remove the return value and instead return 0?

nbagnard commented on odbc/src/api/functions.rs:
```diff
@@ -1785,11 +1785,441 @@ pub unsafe extern "C" fn SQLGetInfo(
+        InfoType::SqlConformance => {
+            // MongoSQL is SQL-92 Entry level compliant.
+            return i16_len::set_output_fixed_data(
+                &SQL_SC_SQL92_ENTRY,
```
Yeah, I don't know if we an Entry level or Intermediate.
We are strictly Entry level because we support case expression (6.9, 2.a) and cast specification (6.10 2.a).
But we might not be strictly Intermediate level either, I did not have time to go over all the restriction on the SQL spec.

mattChiaravalloti replied:
Left it as Entry for now still.

nbagnard commented on odbc/src/api/functions.rs:
```diff
@@ -1785,11 +1785,441 @@ pub unsafe extern "C" fn SQLGetInfo(
+        }
+        // SQL_ODBC_INTERFACE_CONFORMANCE
+        InfoType::OdbcInterfaceConformance => {
+            // The MongoSQL ODBC Driver is currently meets the minimum compliance level.
```
[nit] The sentence is odd.
Did you mean 
```suggestion
            // The MongoSQL ODBC Driver currently meets the minimum compliance level.
```

mattChiaravalloti replied:
good catch, fixed.

nbagnard commented on odbc/src/api/functions.rs:
```diff
@@ -1785,11 +1785,441 @@ pub unsafe extern "C" fn SQLGetInfo(
+        }
+        // SQL_SQL92_PREDICATES
+        InfoType::Sql92Predicates => {
+            // MongoSQL supports the following SQL-92 predicate operators.
```
We also support DISTINCT and EXISTS, so I would add SQL_SP_UNIQUE and SQL_SP_EXISTS, no?

mattChiaravalloti replied:
What is the specific SQL-92 predicate expression that `SQL_SP_UNIQUE` corresponds to? I cannot think of any predicates we have in MongoSQL that would match that.

Also, I have `SQL_SP_EXISTS` first in the list!

nbagnard replied:
ah, I think you are right here :) All good to go 👍

nbagnard commented on odbc/src/api/functions.rs:
```diff
@@ -1785,11 +1785,441 @@ pub unsafe extern "C" fn SQLGetInfo(
+                | SQL_AF_MAX
+                | SQL_AF_MIN
+                | SQL_AF_SUM
+                | SQL_AF_DISTINCT
```
I don't know if we are supposed to return SQL_AF_DISTINCT because we don't support `select Distinct col1, col2 from table1` for example. I couldn't find a description of what SQL_AF_DISTINCT stands for however, so I am not 100% sure.

mattChiaravalloti replied:
My impression was that `SQL_AF_DISTINCT` (standing for "SQL AGGREGATE FUNCTION DISTINCT") corresponded to the `DISTINCT` keyword being used in aggregate functions -- which we do support. The `SQL_AF_*` values all correspond to aggregation functions (`SUM`, `MAX`, etc.), except for two: `SQL_AF_ALL` and `SQL_AF_DISTINCT`. This is why I assumed they were modifiers for aggregation functions.

I doubt it is a meant to indicate `DISTINCT` as in `SELECT DISTINCT`, but I suppose it is possible! Do you know where I can find documentation on the header file values? The files themselves have no documentation, and the docs at `https://learn.microsoft.com/en-us/sql/odbc/reference/...` don't always specifically spell out exactly what each one means.

nbagnard replied:
Yes, this is my problem here too.
I don't know what SQL_AF_DISTINCT stands for and couldn't find the answer. Couldn't find anything useful in the spec and the .h file has no comment :(
I think you can read it both ways, that's why I am very unsure.   
When using  `SELECT <set quantifier> <select list> <table expression>`, one could say that the aggregation function is implicit since it will filter duplicates.
I think you can leave this as is and see how it goes when doing integration with different tools. 
Maybe add a comment saying that we don't know if this is supposed to mean that DISTINCT and ALL are supported as set quantifier for aggregation function or if they can be used as "standalone" set quantifier?

mattChiaravalloti replied:
Ok, I left it for now, then. We can remove it if it causes an issue.

nbagnard commented on odbc/src/api/functions.rs:
```diff
@@ -1785,11 +1785,441 @@ pub unsafe extern "C" fn SQLGetInfo(
+            return i16_len::set_output_fixed_data(&SQL_U16_ZERO, info_value_ptr, string_length_ptr);
+        }
+        // SQL_API_SQLFETCHSCROLL
+        InfoType::ApiSqlFetchScroll => {
```
This is a function id for https://learn.microsoft.com/en-us/sql/odbc/reference/syntax/sqlgetfunctions-function, that's why it is not documented for SQLGetInfo. 
It an error on my part copy-pasting the list of properties from the extracted flow.
You can remove it :)

mattChiaravalloti replied:
Removed! thanks

nbagnard reviewed (changes requested):
Hopefully, I answered all the TODO and did not miss any.
There are also a couple of attributes I am not sure about, like the SQL-92 conformance level, I tried to explain my understanding, train of thought so that it makes sense to you too. We can discuss them if/when needed.

pmeredit reviewed (commented):
Here's some initial comments, nothing needs to be changed yet!

mattChiaravalloti commented on odbc/src/api/functions.rs:
```diff
@@ -1785,11 +1785,441 @@ pub unsafe extern "C" fn SQLGetInfo(
+            // MongoSQL supports that could accept a search pattern. Therefore, the
+            // empty string is always returned.
+            return i16_len::set_output_wstring(
+                "",
```
Hmm, I'm not sure what the docs mean when they say:
> This InfoType is limited to catalog functions. For a description of the use of the escape character in search pattern strings, see [Pattern Value Arguments](https://learn.microsoft.com/en-us/sql/odbc/reference/develop-app/pattern-value-arguments?view=sql-server-ver16).

My interpretation of "catalog functions" here was functions that appear _within_ SQL queries. You are suggesting this applies to the ODBC functions just as "list tables" and "list columns" etc. It's confusing since the docs refer to the `LIKE` operator (which appears in a SQL query, not in the ODBC listing functions). Can you clarify how this InfoType is used?

nbagnard replied:
Yes, this time, I am sure that catalog functions are SQLTables,  SQLColumns, etc... :)
Even, better, this is in the spec 🎉 
See [here](https://learn.microsoft.com/en-us/sql/odbc/reference/develop-app/catalog-functions-in-odbc)
Just for future reference, it might help to remember what they mean by catalog functions : Catalog in ODBC is generally similar to DatabaseMetadata in JDBC. It's everything that is useful for a user to know the DB structure/data model.

mattChiaravalloti replied:
> Catalog in ODBC is generally similar to DatabaseMetadata in JDBC. It's everything that is useful for a user to know the DB structure/data model.

This is super useful for future reference, thank you!

mattChiaravalloti replied:
updated the comment

mattChiaravalloti commented on odbc/src/api/functions.rs:
```diff
@@ -1785,11 +1785,441 @@ pub unsafe extern "C" fn SQLGetInfo(
+            // versions. They use the same numeric value.
+            //
+            // MongoSQL does not have a concept of a "schema" which is a
+            // level above "catalog" (or "database") in the data hierarchy.
```
Updated the comment to (hopefully) more clearly explain why we don't support this!

mattChiaravalloti replied:
done

mattChiaravalloti commented on odbc/src/api/functions.rs:
```diff
@@ -1785,11 +1785,441 @@ pub unsafe extern "C" fn SQLGetInfo(
+            // MongoSQL technically uses the term "database" to describe the
+            // entity described by ODBC's "catalog" term. However, the ODBC
+            // documentation suggests SQL-92 compliant drivers always return
+            // "catalog" for this, so that is what is returned here.
```
Good catch! Changing this one to "database".

mattChiaravalloti replied:
done

mattChiaravalloti commented on odbc/src/api/functions.rs:
```diff
@@ -1785,11 +1785,441 @@ pub unsafe extern "C" fn SQLGetInfo(
+        }
+        // SQL_CATALOG_USAGE
+        InfoType::CatalogUsage => {
+            // TODO: (code review) there is no way to indicate this is read-only. The entire DML bitmask
```
Thanks. Updated the comment and added support for SQL_DATA_SOURCE_READ_ONLY so we return "Y" for that.

mattChiaravalloti commented on odbc/src/api/functions.rs:
```diff
@@ -1785,11 +1785,441 @@ pub unsafe extern "C" fn SQLGetInfo(
+            );
+        }
+        // SQL_GROUP_BY
+        InfoType::GroupBy => {}
```
done. Don't know how I let this one slip by initially 😅

mattChiaravalloti commented on odbc/src/api/functions.rs:
```diff
@@ -1785,11 +1785,441 @@ pub unsafe extern "C" fn SQLGetInfo(
+            return i16_len::set_output_fixed_data(&SQL_U32_ZERO, info_value_ptr, string_length_ptr);
+        }
+        // SQL_GETDATA_EXTENSIONS
+        InfoType::GetDataExtensions => {}
```
done. Don't know how I let this one slip by initially 😅

mattChiaravalloti commented on odbc/src/api/functions.rs:
```diff
@@ -1785,11 +1785,441 @@ pub unsafe extern "C" fn SQLGetInfo(
+        InfoType::TimedateFunctions => {
+            // MongoSQL supports the following timedate functions.
+            const TIMEDATE_FUNCTIONS: u32 = SQL_FN_TD_TIMESTAMPADD // we call it DATEADD
+                | SQL_FN_TD_TIMESTAMPDIFF // we call it DATEDIFF
```
done

mattChiaravalloti commented on odbc/src/api/functions.rs:
```diff
@@ -1785,11 +1785,441 @@ pub unsafe extern "C" fn SQLGetInfo(
+        InfoType::StringFunctions => {
+            // MongoSQL supports the following string functions.
+            const STRING_FUNCTIONS: u32 = SQL_FN_STR_CONCAT
+                | SQL_FN_STR_LTRIM // we actually only support TRIM
```
done

mattChiaravalloti commented on odbc/src/api/functions.rs:
```diff
@@ -1785,11 +1785,441 @@ pub unsafe extern "C" fn SQLGetInfo(
+            // MongoSQL supports ` and " as identifier delimiters. The "
+            // character is the SQL-92 standard, so it is always returned.
+            return i16_len::set_output_wstring(
+                "\"",
```
done

mattChiaravalloti commented on odbc/src/api/functions.rs:
```diff
@@ -1785,11 +1785,441 @@ pub unsafe extern "C" fn SQLGetInfo(
+            //       - MongoDB? Atlas Data Federation? Atlas SQL?
+        }
+        // SQL_DBMS_VER
+        InfoType::DbmsVer => {
```
still working on this! a core package function is a really good idea

mattChiaravalloti replied:
done

mattChiaravalloti commented on odbc/src/api/functions.rs:
```diff
@@ -1785,11 +1785,441 @@ pub unsafe extern "C" fn SQLGetInfo(
+        }
+        // SQL_DBMS_NAME
+        InfoType::DbmsName => {
+            // TODO: (code review) what "name of the DBMS product accessed by the driver" do we want to use?
```
done

mattChiaravalloti commented on odbc/src/api/functions.rs:
```diff
@@ -1785,11 +1785,441 @@ pub unsafe extern "C" fn SQLGetInfo(
+        }
+        // SQL_DRIVER_ODBC_VER
+        InfoType::DriverOdbcVer => {
+            // TODO: (code review) should we use env.odbc_version for this?
```
done

mattChiaravalloti commented on odbc/src/api/functions.rs:
```diff
@@ -1785,11 +1785,441 @@ pub unsafe extern "C" fn SQLGetInfo(
+            // The driver version can be obtained from the Cargo.toml file.
+            // The env! macro call below gets the version from the Cargo file
+            // at compile time.
+            let version = env!("CARGO_PKG_VERSION");
```
done

mattChiaravalloti commented:
@nbagnard I've addressed all of your comments either with code changes or with follow-up questions. There is still some more work to do on this PR: writing the core package function for getting version and unit testing some of the helper functions. It is up to you if you want to look again now since there are a lot of comments or if you want to wait until those last two things are pushed (should be early tomorrow).

mattChiaravalloti commented on odbc/src/api/functions.rs:
```diff
@@ -1785,11 +1785,441 @@ pub unsafe extern "C" fn SQLGetInfo(
+                | SQL_FN_TD_EXTRACT;
+            return i16_len::set_output_fixed_data(&TIMEDATE_FUNCTIONS, info_value_ptr, string_length_ptr);
+        }
+        // SQL_CONVERT_BIGINT
```
removed these

nbagnard reviewed (approved):
LGTM! 💯 
Awesome job! Thank you for going through all of these, even the confusing ones and trying to make sense of them :)

pmeredit commented on odbc/src/api/functions.rs:
```diff
@@ -1794,12 +1794,468 @@ pub unsafe extern "C" fn SQLGetInfo(
+        }
+        // SQL_MAX_IDENTIFIER_LEN
+        InfoType::MaxIdentifierLen => {
+            // MongoSQL does not have a maximum identifier length.
```
Debatable. We can't have an identifier longer than max document size :)

mattChiaravalloti replied:
Ha! True 😄

pmeredit reviewed (approved):
LGTM

mattChiaravalloti commented:
evergreen merge

mattChiaravalloti commented:
evergreen merge

mattChiaravalloti commented:
evergreen merge
